### PR TITLE
ci: build android 32 bit apk

### PR DESCRIPTION
**Pull request by johnny9**
[Windows](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/win64/unsecure_win_gui.zip?branch=pull/317)
[Intel macOS](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/macos/unsecure_mac_gui.zip?branch=pull/317)
[Apple Silicon macOS](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/macos_arm64/unsecure_mac_arm64_gui.zip?branch=pull/317)
[ARM64 Android](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/android/unsecure_android_apk.zip?branch=pull/317)
[ARM32 Android](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/android32/unsecure_android_32bit_apk.zip?branch=pull/317)

### CODE REVIEW

**Comment by jarolrod** [comment]
cc @hebasto

**Comment by johnny9** [comment]
We've had a few people already including myself with armeabi v7a devices. I guess they aren't as old as I thought. Some 2018/2019 devices

**Review by hebasto — COMMENTED** [review]
Approach ACK 9afa997b10250884ffc6261d0babacc680c40f60.

Mind adding a new download link to `.github/PULL_REQUEST_TEMPLATE.md` and `src/qml/README.md`?

**Comment by johnny9** [comment]
from[ 9afa997](https://github.com/bitcoin-core/gui-qml/commit/9afa997b10250884ffc6261d0babacc680c40f60) to [ 14a60c](https://github.com/bitcoin-core/gui-qml/commit/14a60cab5148f647f5be7c1af9c0aa083e088857)
 - Add ARM32 Android link to README
 - Give android32 apk a unique filename

File: src/Makefile.qt.include [hunk 1]
```diff
@@ -518,7 +518,7 @@ QT_BASE_TLD = $(shell tar tf $(QT_BASE_PATH) --exclude='*/*')
 
 bitcoin_qt_apk: FORCE
 	mkdir -p $(APK_LIB_DIR)
-	cp $(dir $(lastword $(CC)))../sysroot/usr/lib/$(host_alias)/libc++_shared.so $(APK_LIB_DIR)
+	cp $(dir $(lastword $(CC)))../sysroot/usr/lib/$(NDK_TOOLCHAIN_PATH)/libc++_shared.so $(APK_LIB_DIR)
```
[the remark this answers is not in the document]
Comment (johnny9) [reply src/Makefile.qt.include:521]: I'll change it to NDK_LIBCXX_DIR.

**Comment by johnny9** [comment]
from[ 14a60ca ](https://github.com/bitcoin-core/gui-qml/commit/14a60cab5148f647f5be7c1af9c0aa083e088857)to[ 6e92cf](https://github.com/bitcoin-core/gui-qml/commit/6e92cf3146f2094f94a2e59e2c1e582ffd0ea6e2)
 - Rename new configure variable to NDK_LIBCXX_DIR

File: .cirrus.yml [hunk 2]
```diff
@@ -386,3 +386,24 @@ task:
...
+  depends_sources_cache:
+    folder: "depends/sources"
+    fingerprint_script: git rev-parse HEAD:depends/packages
+  << : *MAIN_TEMPLATE
+  env:
+    << : *CIRRUS_EPHEMERAL_WORKER_TEMPLATE_ENV
+    FILE_ENV: "./ci/test/00_setup_env_android32.sh"
+  copy_artifacts_script:
+    - cp src/qt/android/build/outputs/apk/debug/android-debug.apk ${CIRRUS_WORKING_DIR}/unsecure_android_32bit.apk
+  unsecure_android_apk_32bit_artifacts:
```
Comment (johnny9) [inline .cirrus.yml:408]: @hebasto will this change the url for the .zip to unsecure_android_apk_32bit.zip?

Comment (hebasto) [reply .cirrus.yml:408]: > @hebasto will this change the url for the .zip to unsecure_android_apk_32bit.zip?

It does. See https://cirrus-ci.org/guide/writing-tasks/#latest-build-artifacts.

We use the following URL format:
```
https://api.cirrus-ci.com/v1/artifact/github/<USER OR ORGANIZATION>/<REPOSITORY>/<TASK NAME OR ALIAS>/<ARTIFACTS_NAME>.zip
```

**Review by hebasto — COMMENTED** [review]
Approach ACK 0face6536d90ea3f4de3019bf60ae7300dee546f.

> Rename new configure variable to NDK_LIBCXX_DIR

Please move this change into the first commit.

> Add ARM32 Android link to README

A similar [change](https://github.com/bitcoin-core/gui-qml/pull/317#pullrequestreview-1386774712) to `.github/PULL_REQUEST_TEMPLATE.md`?

**Comment by hebasto** [comment]
CI is green. An artifact is available to download at https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/android32/unsecure_android_apk_32bit.zip?branch=pull/317.

**Comment by johnny9** [comment]
from[ 0face65 ](https://github.com/bitcoin-core/gui-qml/commit/0face6536d90ea3f4de3019bf60ae7300dee546f)to[ e11b561](https://github.com/bitcoin-core/gui-qml/commit/e11b561d56b447254b813e9e3193fa5d8114dac6)
 - Put change to LIBCXX configure variable to the right commit 
 - Added ARM32 Android url to pull request README

**Comment by jarolrod** [comment]
concept ack, I don't currently have a 32 bit android device so cannot test

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -49,4 +49,5 @@ Links for Windows and macOS build artifacts. Replace <PR> with the assigned pull
 [![Intel macOS](https://img.shields.io/badge/OS-Intel%20macOS-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/macos/unsecure_mac_gui.zip?branch=pull/<PR>)
 [![Apple Silicon macOS](https://img.shields.io/badge/OS-Apple%20Silicon%20macOS-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/macos_arm64/unsecure_mac_arm64_gui.zip?branch=pull/<PR>)
 [![ARM64 Android](https://img.shields.io/badge/OS-Android-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/android/unsecure_android_apk.zip?branch=pull/<PR>)
+[![ARM32 Android](https://img.shields.io/badge/OS-Android%2032bit-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/android32/unsecure_android_32bit_apk.zip?branch=pull/<PR>)
```
Comment (hebasto) [inline .github/PULL_REQUEST_TEMPLATE.md:52]: https://github.com/bitcoin-core/gui-qml/pull/317/files#r1167988459:

```suggestion
[![ARM32 Android](https://img.shields.io/badge/OS-Android%2032bit-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/android32/unsecure_android_apk_32bit.zip?branch=pull/<PR>)
```

**Review by hebasto — COMMENTED** [review]
The artifact name and urls are not inlined now. Mind updating urls or an artifact name s/`unsecure_android_apk_32bit_artifacts`/`unsecure_android_32bit_apk_artifacts`/?

**Comment by 0xB10C** [comment]
> CI is green. An artifact is available to download at https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/android32/unsecure_android_apk_32bit.zip?branch=pull/317.

Synced signet on a OnePlus One (2014, 3GB RAM,  4x 2.45 GHz Snapdragon 801) with this APK in about an hour.

**Review by jarolrod — APPROVED** [review]
ACK e11b561d56b447254b813e9e3193fa5d8114dac6

I did not test this and don't currently have a way to test. But we have confirmation of two users being able to install and run, and one user being able to sync.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 3 above)
Comment (jarolrod) [inline .github/PULL_REQUEST_TEMPLATE.md:52]: @johnny9 this should fix the issue of not being able to grab the link from the PR description

**Comment by johnny9** [comment]
from[ e11b561 ](https://github.com/bitcoin-core/gui-qml/commit/e11b561d56b447254b813e9e3193fa5d8114dac6)to[ ef](https://github.com/bitcoin-core/gui-qml/commit/efc4dd3a2dba34b0823b53ded51b9386917d5344)
 - fixed the url in the pull request template

**Comment by johnny9** [comment]
from e11b561 to  3b5068
 - Changed the cirrus task artifact property to unsecure_android_32bit_apk_artifacts to match the filename of the artifact and match the url

**Comment by johnny9** [comment]
from[ 3b50687 ](https://github.com/bitcoin-core/gui-qml/commit/3b5068719066b1a3d42207ac4493c51b53e37543)to[ ec8e89](https://github.com/bitcoin-core/gui-qml/commit/ec8e8969f6749ab68f72a9db841212754c78077a)
  - rebased with main

**Comment by hebasto** [comment]
@johnny9 
> from[ 3b50687 ](https://github.com/bitcoin-core/gui-qml/commit/3b5068719066b1a3d42207ac4493c51b53e37543)to[ ec8e89](https://github.com/bitcoin-core/gui-qml/commit/ec8e8969f6749ab68f72a9db841212754c78077a)
> 
>    * rebased with main

Thanks for rebasing!

Although another diff is required after rebasing:
```diff
--- a/.cirrus.yml
+++ b/.cirrus.yml
@@ -399,7 +399,9 @@ task:
   alias: android32
   << : *CONTAINER_DEPENDS_TEMPLATE
   container:
-    image: ubuntu:jammy
+      CI_IMAGE_NAME_TAG: ubuntu:jammy
+      FILE_ENV: "./ci/test/00_setup_env_android32.sh"
+  << : *CREDITS_TEMPLATE
   android_sdk_cache:
     folder: "depends/SDKs/android"
     fingerprint_key: "ANDROID_API_LEVEL=28 ANDROID_BUILD_TOOLS_VERSION=28.0.3 ANDROID_NDK_VERSION=23.2.8568313"
@@ -409,7 +411,6 @@ task:
   << : *MAIN_TEMPLATE
   env:
     << : *CIRRUS_EPHEMERAL_WORKER_TEMPLATE_ENV
-    FILE_ENV: "./ci/test/00_setup_env_android32.sh"
   copy_artifacts_script:
     - cp src/qt/android/build/outputs/apk/debug/android-debug.apk ${CIRRUS_WORKING_DIR}/unsecure_android_32bit.apk
   unsecure_android_32bit_apk_artifacts:
```

**Review by hebasto — APPROVED** [review]
ACK 28ab1ee7e5d158804d3c14b1b3e2a1fe0241c115

**Pull request merged**